[MTD simulation, validation]: implementation of MTD clusters association maps

martinamalberti opened the pull request:
#### PR description:
This pull request implements association maps for MTD reco clusters to MtdSimLayerClusters and MtdSimLayerClusters to Tracking Particles. A summary was given at the last MTD DPG meeting [1].

The MTD clusters associations maps are implemented in the following packages:
SimDataFormats/Associations
SimFastTiming/MtdAssociationProducers
and used in the Validation/MtdValidation package.

Some changes were made also to the following packages:
- SimDataFormats/CaloAnalysis: addition of new members (positions_, idOffset_ ) and methods to the MtdSimCluster class. Class version updated accordingly.
- SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc updated to:  
   1) produce all the MTD sim clusters, not only those from direct hits in the MTD, relying on the BTL sim hits categorization implemented by F. Cossutti in https://github.com/cms-sw/cmssw/blob/master/SimG4CMS/Forward/interface/MtdSD.h;   
   2) store the local position of the MTD sim cluster estimated from the true position of the sim hits.
- Geometry/MTDGeometryBuilder: added a new method to the MTDGeomUtil class to return the sensor module DetId starting from the crystal BTLDetId. 

[1] https://indico.cern.ch/event/1376660/#6-mtd-clusters-association-map

#### PR validation:
The code was tested on the workflows:
24807.0 (SingleMuPt10)
24906.0 (SinglePiFlatPt0p7To10) 
and produces the expected results.

martinamalberti committed:
implementation of MTD clusters association maps

martinamalberti committed:
set debug option to false

martinamalberti committed:
fix histogram names

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"030c808dd7f5a474221a06866de63cf387832ea1":{"files":["Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc"],"squashed":false,"time":1707392309},"3e2c7e7a76e145c35a36571793c61157979a11f0":{"files":["SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc"],"squashed":false,"time":1707388531},"dbe659978fe36ccfdd8df0e43cba4c406185b714":{"files":["Geometry/MTDGeometryBuilder/interface/MTDGeomUtil.h","Geometry/MTDGeometryBuilder/src/MTDGeomUtil.cc","SimDataFormats/Associations/interface/MtdRecoClusterToSimLayerClusterAssociationMap.h","SimDataFormats/Associations/interface/MtdRecoClusterToSimLayerClusterAssociator.h","SimDataFormats/Associations/interface/MtdRecoClusterToSimLayerClusterAssociatorBaseImpl.h","SimDataFormats/Associations/interface/MtdSimLayerClusterToRecoClusterAssociationMap.h","SimDataFormats/Associations/interface/MtdSimLayerClusterToTPAssociator.h","SimDataFormats/Associations/interface/MtdSimLayerClusterToTPAssociatorBaseImpl.h","SimDataFormats/Associations/src/MtdRecoClusterToSimLayerClusterAssociationMap.cc","SimDataFormats/Associations/src/MtdRecoClusterToSimLayerClusterAssociator.cc","SimDataFormats/Associations/src/MtdRecoClusterToSimLayerClusterAssociatorBaseImpl.cc","SimDataFormats/Associations/src/MtdSimLayerClusterToRecoClusterAssociationMap.cc","SimDataFormats/Associations/src/MtdSimLayerClusterToTPAssociator.cc","SimDataFormats/Associations/src/MtdSimLayerClusterToTPAssociatorBaseImpl.cc","SimDataFormats/Associations/src/classes.h","SimDataFormats/Associations/src/classes_def.xml","SimDataFormats/CaloAnalysis/interface/MtdSimCluster.h","SimDataFormats/CaloAnalysis/interface/MtdSimLayerClusterFwd.h","SimDataFormats/CaloAnalysis/src/classes_def.xml","SimFastTiming/MtdAssociatorProducers/plugins/BuildFile.xml","SimFastTiming/MtdAssociatorProducers/plugins/MtdRecoClusterToSimLayerClusterAssociatorByHitsImpl.cc","SimFastTiming/MtdAssociatorProducers/plugins/MtdRecoClusterToSimLayerClusterAssociatorByHitsImpl.h","SimFastTiming/MtdAssociatorProducers/plugins/MtdRecoClusterToSimLayerClusterAssociatorByHitsProducer.cc","SimFastTiming/MtdAssociatorProducers/plugins/MtdRecoClusterToSimLayerClusterAssociatorEDProducer.cc","SimFastTiming/MtdAssociatorProducers/plugins/MtdSimLayerClusterToTPAssociatorByTrackIdImpl.cc","SimFastTiming/MtdAssociatorProducers/plugins/MtdSimLayerClusterToTPAssociatorByTrackIdImpl.h","SimFastTiming/MtdAssociatorProducers/plugins/MtdSimLayerClusterToTPAssociatorByTrackIdProducer.cc","SimFastTiming/MtdAssociatorProducers/plugins/MtdSimLayerClusterToTPAssociatorEDProducer.cc","SimFastTiming/MtdAssociatorProducers/python/mtdRecoClusterToSimLayerClusterAssociation_cfi.py","SimFastTiming/MtdAssociatorProducers/python/mtdSimLayerClusterToTPAssociation_cfi.py","SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc","Validation/Configuration/python/mtdSimValid_cff.py","Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc","Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc","Validation/MtdValidation/plugins/MtdTracksValidation.cc"],"squashed":false,"time":1707387043}},"emoji":{},"last_seen_sha":"030c808dd7f5a474221a06866de63cf387832ea1","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43913/38763

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/Associations/src/classes_def.xml  modified in PR(s): #37507
  - File  SimDataFormats/CaloAnalysis/src/classes_def.xml  modified in PR(s): #43468

fabiocos commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37305/summary.html
**COMMIT**: 030c808dd7f5a474221a06866de63cf387832ea1
**CMSSW**: CMSSW_14_1_X_2024-02-08-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43913/37305/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 87 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-08-1100+c21e5b/60965/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 59 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-08-1100+c21e5b/60965/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248626
 - DQMHistoTests: [Total failures: 1210](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-08-1100+c21e5b/60965/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247394
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 119.90400000000002 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.777 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-08-1100+c21e5b/60965/triggerResults): no differences found

martinamalberti committed:
added histogram of BTL clus categorization

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43913/38780

- This PR adds an extra 76KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/Associations/src/classes_def.xml  modified in PR(s): #37507
  - File  SimDataFormats/CaloAnalysis/src/classes_def.xml  modified in PR(s): #43468
  - File  Validation/MtdValidation/plugins/MtdTracksValidation.cc  modified in PR(s): #43918

cmsbuild commented:
Pull request #43913 was updated. @Dr15Jones, @srimanob, @nothingface0, @mdhildreth, @subirsarkar, @makortel, @bsunanda, @civanch, @rvenditti, @antoniovagnerini, @tjavaid, @cmsbuild, @syuvivida can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37325/summary.html
**COMMIT**: 8d0fb224b56a57176ef28d1e5400c41f08a17c6d
**CMSSW**: CMSSW_14_1_X_2024-02-09-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43913/37325/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-09-1100+c21e5b/60984/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 38 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-09-1100+c21e5b/60984/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248626
 - DQMHistoTests: [Total failures: 24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-09-1100+c21e5b/60984/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3248580
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 120.264 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-09-1100+c21e5b/60984/triggerResults): no differences found

martinamalberti commented:
Changes in the MTD/Tracks plots are expected because of this modification in MtdTracksValidation.cc:
https://github.com/cms-sw/cmssw/pull/43913/files#diff-035207001dc7d8a415f150209e67549137f2abeee744f4a55af9d9620b691babL615

srimanob commented:
type mtd

martinamalberti commented:
@civanch 
 
> 1. is it correct to use type size_t? in other places we use to use std::size_t

To my understanding the two should be almost equivalent. I see occurrences of both in many places in cmssw. If there is any prescription in either way I will be happy to follow it.

> 2. you add 3 small classes implementing special maps used in your project. Why not to use std::map or std::unordered_map instead?

We agreed with @fabiocos  to implement these associations for MTD using a structure of code similar to what was already done for other subsystems, like HGCal and Tracker. See for example:

https://github.com/cms-sw/cmssw/tree/master/SimDataFormats/Associations/
https://github.com/cms-sw/cmssw/tree/master/SimCalorimetry/HGCalAssociatorProducers
https://github.com/cms-sw/cmssw/tree/master/SimTracker/TrackAssociation
https://github.com/cms-sw/cmssw/tree/master/SimTracker/TrackAssociatorProducers

For the case of association maps between MTD FTLCLusters and MtdSimLayerClusters we introduced 2 additional small classes to overcome issues with nested edm:Refs met when dealing with the FTLClusterCollection (which is a DetSetVec of FTLClusters), not handled by standard cmssw AssociationMaps.

The implementation of those 2 classes is inspired to what is done, for example, for the tracker here:
https://github.com/cms-sw/cmssw/blob/master/SimTracker/TrackerHitAssociation/interface/ClusterTPAssociation.h

civanch commented:
+1

OK to me

makortel commented:
> We agreed with @fabiocos to implement these associations for MTD using a structure of code similar to what was already done for other subsystems, like HGCal and Tracker.

AFAICT both HGCal and Tracker use `edm::AssociationMap<edm::OneToManyWithQualityGeneric<...>>` 
https://github.com/cms-sw/cmssw/blob/3a4a13be92b40a14b4211c7f267180e1cce4f711/SimDataFormats/Associations/interface/LayerClusterToCaloParticleAssociatorBaseImpl.h#L21-L26
https://github.com/cms-sw/cmssw/blob/3a4a13be92b40a14b4211c7f267180e1cce4f711/SimDataFormats/Associations/interface/TrackAssociation.h#L16-L22
that internally uses `std::map`. 

A sorted `std::vector` of `std::pair` is, however, generally better for "write once, search many times" and IO.

Here, if I understood correctly, the contained `vector`s in `MtdRecoClusterToSimLayerClusterAssociationMap` and `MtdSimLayerClusterToRecoClusterAssociationMap` are not sorted (given the linear search e.g. in
https://github.com/cms-sw/cmssw/blob/8d0fb224b56a57176ef28d1e5400c41f08a17c6d/Validation/MtdValidation/plugins/BtlLocalRecoValidation.cc#L665-L668

If a custom data structure is used for the associations, in my opinion that data structure should also provide an interface for the search, and if implemented with `vector`s, sort the `vector`s (or would the number of elements be so low that a linear search would be faster than a binary search?).

makortel commented on SimDataFormats/Associations/src/classes.h:
```diff
@@ -39,12 +43,26 @@ namespace SimDataFormats_Associations {
+    MtdSimLayerClusterToRecoClusterAssociationMap dummy19;
+    edm::Wrapper<MtdSimLayerClusterToRecoClusterAssociationMap> dummy20;
+
+    edm::Wrapper<reco::MtdSimLayerClusterToTPAssociator> dummy21;
```
This entire `namespace SimDataFormats_Associations { ... }` has not been needed for a long time. Maybe we could use this PR to remove it?

makortel commented on SimDataFormats/Associations/src/classes_def.xml:
```diff
@@ -141,4 +141,39 @@
+
+
+  
+  <class name="std::pair<FTLClusterRef, std::vector<MtdSimLayerClusterRef> >" persistent="true" />
```
While `persistent="true"` would technically be ok, I'd suggest to remove it and rely it being the default. This would make it easier to focus the reader's eye on the `persistent="false"` cases.

makortel commented on SimDataFormats/Associations/src/classes_def.xml:
```diff
@@ -141,4 +141,39 @@
+  
+  <class name="std::pair<FTLClusterRef, std::vector<MtdSimLayerClusterRef> >" persistent="true" />
+  <class name="std::vector<std::pair<FTLClusterRef, std::vector<MtdSimLayerClusterRef> > >" />
+  <class name="MtdRecoClusterToSimLayerClusterAssociationMap" persistent="true" />
```
Needs a class version

makortel commented on SimDataFormats/Associations/src/classes_def.xml:
```diff
@@ -141,4 +141,39 @@
+  <class name="edm::Wrapper<MtdRecoClusterToSimLayerClusterAssociationMap>" />
+  <class name="std::pair<MtdSimLayerClusterRef, std::vector<FTLClusterRef> >" persistent="true" />
+  <class name="std::vector<std::pair<MtdSimLayerClusterRef, std::vector<FTLClusterRef> > >" />
+  <class name="MtdSimLayerClusterToRecoClusterAssociationMap" persistent="true" />
```
Needs a class version.

martinamalberti committed:
update with cluster sorting in mtd assoc maps

cmsbuild commented:
Pull request #43913 was updated. @antoniovagnerini, @srimanob, @nothingface0, @tjavaid, @rvenditti, @mdhildreth, @cmsbuild, @subirsarkar, @syuvivida, @civanch can you please check and sign again.

civanch commented:
please test

fabiocos commented on SimDataFormats/Associations/interface/MtdSimLayerClusterToRecoClusterAssociationMap.h:
```diff
@@ -39,9 +42,31 @@ class MtdSimLayerClusterToRecoClusterAssociationMap {
 private:
+  static bool compare(const value_type& i, const value_type& j) {
+    auto i_hAndE = (i.first)->hits_and_energies();
+    std::vector<uint64_t> i_ids(i_hAndE.size());
```
@martinamalberti do we really need to go through the creation of these temporary vectors, that might be expensive, instead of defining a comparison function for pairs that just compare the ```.first``` element?

martinamalberti replied:
@fabiocos Thanks. Changed here:
https://github.com/martinamalberti/cmssw/blob/mm_MTDAssociationMaps/SimDataFormats/Associations/interface/MtdSimLayerClusterToRecoClusterAssociationMap.h#L53-L64

makortel commented on SimDataFormats/Associations/interface/MtdRecoClusterToSimLayerClusterAssociationMap.h:
```diff
@@ -0,0 +1,57 @@
+    map_.emplace_back(recoClus, simClusVect);
+  }
+
+  void sort() { std::sort(map_.begin(), map_.end(), compare); }
```
If the function would be called `post_insert()`, the framework would call it automatically when the data product is put in the Event, that would allow to not to rely on the developer to remember to call the `sort()` in the EDProducer.

martinamalberti replied:
@makortel. Thank you. PR updated to use post_insert()

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37424/summary.html
**COMMIT**: 11f4db634e6c6c2c3936947074f3c662a4212881
**CMSSW**: CMSSW_14_1_X_2024-02-13-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43913/37424/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-1100+c21e5b/61065/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 55 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-1100+c21e5b/61065/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248554
 - DQMHistoTests: [Total failures: 33](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-1100+c21e5b/61065/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3248499
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 120.264 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-1100+c21e5b/61065/triggerResults): no differences found

martinamalberti committed:
removed unnecessary auxiliary vectors + use post_insert

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43913/38851

- This PR adds an extra 144KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/Associations/src/classes_def.xml  modified in PR(s): #37507
  - File  Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc  modified in PR(s): #43928
  - File  Validation/MtdValidation/plugins/MtdTracksValidation.cc  modified in PR(s): #43918, #43928

cmsbuild commented:
Pull request #43913 was updated. @mdhildreth, @rvenditti, @nothingface0, @antoniovagnerini, @subirsarkar, @cmsbuild, @syuvivida, @civanch, @tjavaid, @srimanob can you please check and sign again.

fabiocos commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37452/summary.html
**COMMIT**: 831672717e4d206aaf4404e4ce9fa1d7acdda654
**CMSSW**: CMSSW_14_1_X_2024-02-13-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43913/37452/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 95 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 46 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248554
 - DQMHistoTests: [Total failures: 24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3248508
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 120.264 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/triggerResults): no differences found

martinamalberti commented:
Results for MTD/Tracks  comparisons are as expected.

> +1
> 
> **Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37452/summary.html **COMMIT**: [8316727](https://github.com/cms-sw/cmssw/commit/831672717e4d206aaf4404e4ce9fa1d7acdda654) **CMSSW**: CMSSW_14_1_X_2024-02-13-2300/el8_amd64_gcc12 **User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43913/37452/install.sh` to create a dev area with all the needed externals and cmssw changes.
> 
> ## Comparison Summary
> **Summary**:
> 
> * You potentially [removed 95 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR/logRootQA-events.log) from the logs
> * [Reco comparison results: 46 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR.html)
> * DQMHistoTests: Total files compared: 48
> * DQMHistoTests: Total histograms compared: 3248554
> * DQMHistoTests: [Total failures: 24](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/validateJR.html)
> * DQMHistoTests: Total nulls: 0
> * DQMHistoTests: Total successes: 3248508
> * DQMHistoTests: Total skipped: 22
> * DQMHistoTests: Total Missing objects: 0
> * DQMHistoSizes: Histogram memory added: 120.264 KiB( 47 files compared)
> * DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
> * DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
> * Checked 200 log files, 161 edm output root files, 48 DQM output files
> * [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-13-2300+c21e5b/61089/triggerResults): no differences found

civanch commented:
+1

fabiocos commented:
@cms-sw/dqm-l2 @cms-sw/upgrade-l2 any comment?

srimanob commented:
Hi @martinamalberti @fabiocos

Is the error report on dictionary checking expected?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37424/dupDict/lostDefs.txt
I understand that you add them, but it should not show error on checking, I assume.

makortel commented:
> Is the error report on dictionary checking expected?
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37424/dupDict/lostDefs.txt

I _think_ the checker complains that (it believes) the dictionaries
https://github.com/cms-sw/cmssw/blob/831672717e4d206aaf4404e4ce9fa1d7acdda654/SimDataFormats/Associations/src/classes_def.xml#L177-L178
would be misplaced. Looking at the dependencies
* `TrackingParticle` is defined in `SimDataFormats/TrackingAnalysis`
* `MtdSimLayerCluster` is defined in `SimDataFormats/CaloAnalysis`
* `SimDataFormats/Associations` already depends on both `TrackingAnalysis` and `CaloAnalysis`

the `SimDataFormats/Associations` looks like a reasonable place for these dictionaries. Then the checker should be updated. I'd extend this line
https://github.com/cms-sw/cmssw/blob/8520d4d557cf162b8990e8cda72e3c3be57a7e91/Utilities/ReleaseScripts/scripts/duplicateReflexLibrarySearch.py#L35
with
```py
, 'MtdSimLayerCluster.+TrackingParticle', 'TrackingParticle.+MtdSimLayerCluster'
```

martinamalberti committed:
update checker

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43913/38954

- This PR adds an extra 148KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  SimDataFormats/Associations/src/classes_def.xml  modified in PR(s): #37507
  - File  Validation/MtdValidation/plugins/EtlLocalRecoValidation.cc  modified in PR(s): #43928
  - File  Validation/MtdValidation/plugins/MtdTracksValidation.cc  modified in PR(s): #43918, #43928

cmsbuild commented:
Pull request #43913 was updated. @antoniovagnerini, @subirsarkar, @rvenditti, @syuvivida, @Dr15Jones, @nothingface0, @smuzaffar, @srimanob, @makortel, @cmsbuild, @tjavaid can you please check and sign again.

srimanob commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37573/summary.html
**COMMIT**: 1cebff77b6cb6bf13c4b0316c4d5e2f668180c4b
**CMSSW**: CMSSW_14_1_X_2024-02-19-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43913/37573/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 100 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-19-2300+c21e5b/61216/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 57 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-19-2300+c21e5b/61216/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248389
 - DQMHistoTests: [Total failures: 1210](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-19-2300+c21e5b/61216/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247157
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 120.264 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-19-2300+c21e5b/61216/triggerResults): no differences found

srimanob commented:
+Upgrade

fabiocos commented:
@cms-sw/core-l2 further comments/suggestions? @cms-sw/dqm-l2 any comment?

martinamalberti commented:
@cms-sw/core-l2 , @cms-sw/dqm-l2: any further comment on this PR?

tjavaid commented:
Hi @martinamalberti and @fabiocos , sorry for a bit late response. However, I cannot access e.g. [Comparison failed](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-19-2300+c21e5b/61216/validateJR.html) (although I can see in the Summary). Is there a need to re-trigger the tests?

fabiocos commented:
please test

@tjavaid the output is no more available after one week...

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c21e5b/37810/summary.html
**COMMIT**: 1cebff77b6cb6bf13c4b0316c4d5e2f668180c4b
**CMSSW**: CMSSW_14_1_X_2024-02-29-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43913/37810/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 18 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-29-1100+c21e5b/61440/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 54 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-29-1100+c21e5b/61440/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3338784
 - DQMHistoTests: [Total failures: 30](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-29-1100+c21e5b/61440/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3338732
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 120.264 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 12.837 KiB       MTD/BTL
 - DQMHistoSizes: changed ( 23234.0,... ): 7.207 KiB        MTD/ETL
 - Checked 205 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-29-1100+c21e5b/61440/triggerResults): no differences found

makortel commented on SimDataFormats/Associations/src/classes.h:
```diff
@@ -15,36 +15,52 @@
+  edm::Wrapper<MtdSimLayerClusterToRecoClusterAssociationMap> dummy20;
+
+  edm::Wrapper<reco::MtdSimLayerClusterToTPAssociator> dummy21;
+};
```
This `struct SimDataFormats_Associations { ... };` could be removed (those have been unnecessary for a long time), but at this stage it could be left for a future PR.

fabiocos replied:
@makortel do you mean the whole block lines 23 - 66? We may add small PR after this one is integrated

makortel replied:
> do you mean the whole block lines 23 - 66?

Yes.

> We may add small PR after this one is integrated

Thanks!

makortel commented:
+core

tjavaid commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @rappoccio, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
